gain instead of Emax, ADC options and more

sagitta42 opened the pull request:
-  small bug fixes in `baselines.jl`
- small bug fixes for fieldgen file caching in `detector.jl` and `legend_detector_to_siggen.jl`: no cached name provided = not caching simulation (for SSD or siggen); SSD works fine, but for siggen I had to create temporary configs in `cache/` folder with keyword `tmp`. The bug was that when a different simulation was launched, it read the "cached" tmp file anyway
- changed the "maximum energy" parameterization of preamp (where Emax is provided and set that `typemax(UInt16) = Emax`, and gain calculated based on that) to providing gain directly in units ADC/eV, and if then the ADC value goes over max of UInt16, it gets clamped. Changes in `preamp.jl`. Changes in other files because now gain is in ADC/eV and not ADC/charge. Will reconsider in the future what's more convenient, for now ADC/eV seems more convenient because can tune gain for example on the FEP peak
- option to provide noise in units of ADC rather than keV (changes in `preamp.jl` and `noise.jl`)
- option to provide offset in ADC rather than keV (in `preamp.jl`)
- Felix's bugfix to remove hits with no energy depositions in `pet_to_stp.jl`
- added option to provide a pre-generated pss table to `pss_to_raw()` (before could only provide the path to the saved pss hdf5 file, now also table in code)
- optional argument `n_waveforms` that allows to simulated a limited amount of waveforms to process from given file (in `stp_to_pss` and `pss_to_raw`), convenient for testing or tutorials

Commit:
gain instead of Emax, ADC options and more

oschulz commented:
Good to merge from my side.

fhagemann commented on src/baselines.jl:
```diff
@@ -91,18 +94,31 @@ end
+    base_start = wf.signal[1] # value of waveform at the beginning
+    peak_index = risepoint(wf) # value of wf right before signal rising i.e. end of baseline
     baseline = wf.signal[begin:peak_index]
+    # is this supposed to remove pileup? i think it is base_slope() due to it not being flat
```
Yes, this seems like a pileup cut on the baseline (making sure that the first baseline value is within the interval `base_lolim` to `base_uplim` and that it has no large slope (such that the mean is somewhat close to the first value)

sagitta42 replied:
Thanks! The "baseline catalog" part of LegendGeSim is underdeveloped for now because it doesn't work for HADES data, as it does not usually contain all-baseline "empty" waveforms, unlike L200 data where in some trigger modes when one detectors triggers all record waveforms. So I imagine we will come back to this later when moving to L200 data simulation. What happens now with the `NoiseFromData` mode, it simply extracts noise and offset from there and adds that to the waveform, which is basically really same as `NoiseFromSim` in that it only adds this high frequency Gaussian noise.

fhagemann commented on src/baselines.jl:
```diff
@@ -91,18 +94,31 @@ end
+        return false
+    end
+
     # wveform value cut 
     cut_value = wf.signal[1000] > 1000
 
     # peak cut 
     peak = findmax(wf.signal)[2]
     cut_peak = peak < 2100 && peak > 1650
```
These seem like hard-coded values. Do they always apply or does it make sense to make these keyword arguments?

sagitta42 replied:
Yeah this was done by a master student in our group years ago, and the values are hard-coded. Not sure if he tuned them for HADES data, or took them from quality cuts at the time. All this should be developed better later for L200.

fhagemann commented on src/detector.jl:
```diff
@@ -42,6 +42,11 @@ Look up fieldgen generated electric field and weighting potential files
 function simulate_detector(det_meta::PropDict, env::Environment, simulator::SiggenSimulator;
         overwrite::Bool = false)
 
+    # if no cached name given, force simulation from scratch (or else might read past "tmp" file)
+    if simulator.cached_name == ""
+        overwrite = true
+    end
```
Maybe add a `warning` or `info` here, that the keyword `overwrite` will be overwritten (?)

sagitta42 replied:
Yeah this part with messages should all change. For now the main @info messages happen in `PSSimulator`, which makes it print the same @info two times confusing the user when SSD geometry is constructed to remove events outside of the detector in `pet_to_stp()` :') So now in `PSSimulator` it says "No cached name provided, not caching the simulation". I'll get on cleaning up the messages at some point.

fhagemann commented on src/noise.jl:
```diff
@@ -153,7 +154,8 @@ function simulate_noise(wf::RDWaveform, preamp::PreAmp)
     # if we're not simulating noise from scratch, we'll do this anyway, but noise sigma will be 0
     # -> maybe there's a better way?
-    noise_σ = ustrip(uconvert(u"eV", preamp.noise_σ))
+    noise_σ = preamp.noise_σ_keV == 0 ? preamp.noise_σ_ADC / preamp.gain : uconvert(u"eV", preamp.noise_σ_keV)
```
According to the [definition in `preamp.jl`](https://github.com/legend-exp/LegendGeSim.jl/pull/49/files#diff-2849d998958e76bf52938428b75e136b23228359a0a267f3724380d8291369baR59), `preamp.noise_σ_keV` always has the unit `u"keV"`.
Therefore, `preamp.noise_σ_keV == 0` will never be true because the `0` is missing a unit.
This could be avoided by using `iszero(preamp.noise_σ_keV)` instead.

sagitta42 replied:
You're right! Eagle eye. Fixed it.

fhagemann commented on src/noise.jl:
```diff
@@ -153,7 +154,8 @@ function simulate_noise(wf::RDWaveform, preamp::PreAmp)
     # -> maybe there's a better way?
-    noise_σ = ustrip(uconvert(u"eV", preamp.noise_σ))
+    noise_σ = preamp.noise_σ_keV == 0 ? preamp.noise_σ_ADC / preamp.gain : uconvert(u"eV", preamp.noise_σ_keV)
+    noise_σ = ustrip(noise_σ)
```
Just to be safe that `noise_σ` is always in units of `eV`, you could do `ustrip(u"eV", noise_σ)` here.

sagitta42 replied:
Thanks, added this to be sure, ran tests with noise in ADC input, works correctly.

fhagemann replied:
It would be good to add the tests you're talking about also to `test/runtest.jl`, such that they get checked with every modification of the source code in the future!

fhagemann commented on src/pet_to_stp.jl:
```diff
@@ -46,6 +46,9 @@ function pet_to_stp(pet_table::Table, detector_SSD::SolidStateDetector)
     hits = RadiationDetectorSignals.ungroup_by_evtno(events_clustered)
 
+    @info "...removing hits with no energy deposits"
+    hits = hits[findall(edep -> !iszero(edep), hits.edep)]
```
Not sure whether it's more efficient to use `filter!` here: `filter!(hits, h -> !iszero(h.edep))`
@oschulz ?

fhagemann commented on src/preamp.jl:
```diff
@@ -25,18 +25,40 @@ This struct is currently mutable because in the case of noise modelling based on
+    # gain = offset == 0u"keV" ? 0 : typemax(UInt16) / ((max_e+offset)/uconvert(u"keV", germanium_ionization_energy))
+    # gain = offset_ADC == 0 ? : ( typemax(UInt16) - offset_ADC ) * uconvert(u"keV", germanium_ionization_energy) / max_e
+
+    "Preamp Gaussian noise in ADC"
+    noise_σ_ADC = 0
+
+    "Preamp Gaussian noise in keV"
+    noise_σ_keV::typeof(1.0*energy_unit)= 0u"keV"
```
It might be helpful to note somewhere that if both values are given, `noise_σ_keV` has priority and will be the value used.

sagitta42 replied:
Yes I made a ToDo comment for myself somewhere to add a check, if both are added to either throw an error or a warning. Same for offset in keV or ADC, big warning if both are provided, both will be double added.

fhagemann commented on src/preamp.jl:
```diff
@@ -97,30 +125,38 @@ RDWaveform, GenericPreAmp -> RDWaveform
+    gain = ustrip(preamp.gain)
+    wf_preamp = RDWaveform(wf_preamp.time, wf_preamp.signal * gain)
+
+    ## ADC offset: if keV offset is used, ADC offset is zero
+    wf_preamp = RDWaveform(wf_preamp.time, wf_preamp.signal .+ preamp.offset_ADC)
```
Where exactly do you ensure that ADC offset is zero if keV offset is used? 
I didn't manage to quickly find it..

sagitta42 replied:
I don't ensure it anywhere because I was lazy :D But since it's now gonna go tutorial level, you're right. Added an error if either noise/offset are provided in both ADC and keV

fhagemann replied:
I mean you could also do something like (DISCLAIMER: not tested)
```julia
if iszero(preamp.offset_keV)
    wf_preamp = RDWaveform(wf_preamp.time, wf_preamp.signal .+ preamp.offset_ADC)
elseif !iszero(preamp.offset_ADC)
     @warn "Both offset_ADC and offset_keV were given. Ignoring offset_ADC"
end
```
This will give some feedback to the user without crashing the code and resulting in an error.

fhagemann commented on src/preamp.jl:
```diff
@@ -97,30 +125,38 @@ RDWaveform, GenericPreAmp -> RDWaveform
     ## gain 
-    wf_preamp = RDWaveform(wf_preamp.time, wf_preamp.signal * preamp.gain)
+    # gain is in eV^-1
+    gain = ustrip(preamp.gain)
```
Ensure correct units by using `ustrip(u"eV", preamp.gain)`

sagitta42 replied:
Shouldn't it be somehow `eV^-1`? If I put `ustrip(u"eV", ...)` it will strip even if it's inverse?..

fhagemann replied:
Yes, you are right. `ustrip(u"eV^-1", ...)` should be correct.

fhagemann commented on src/preamp.jl:
```diff
@@ -97,30 +125,38 @@ RDWaveform, GenericPreAmp -> RDWaveform
     # wf values are in eV but without u"eV" attached
-    offset = ustrip(uconvert(u"eV", preamp.offset))
-    wf_preamp = RDWaveform(wf_preamp.time, wf_preamp.signal .+ offset)
+    offset_keV = ustrip(uconvert(u"eV", preamp.offset_keV))
```
The naming is a bit weird because `offset_keV` now has units of `eV`.

sagitta42 replied:
You're right, fixed it.
Actually ToDo to untangle the unit threads here and see why waveform signal is not `u"eV"` but "unitless eV". Would be better practice to keep the unit.

fhagemann replied:
I agree.

fhagemann commented on src/preamp.jl:
```diff
@@ -25,18 +25,40 @@ This struct is currently mutable because in the case of noise modelling based on
+    #         # typemax(UInt16) / ((max_e+offset_keV)/uconvert(u"keV", germanium_ionization_energy)) # ADC/charge
+    #         typemax(UInt16) / uconvert(u"eV", max_e + offset_keV) # ADC/eV
+    #     end            
+    gain::typeof(1.0/1u"eV")
```
I think this one needs a default value.

sagitta42 replied:
Well, it's different for every preamp, so the default value might be off. But you're right easier to have it for reference, same as tau decay and rise values. I will update them also to correspond more to reality, 15ns risetime is way too small

fhagemann replied:
What I mean is that the way this is implemented in code `@with_kw mutable struct GenericPreamp`
allowing for keyword arguments (`@with_kw`) works best if every field has a default value.
Of course, the default value might be off, but you can always overwrite it afterwards as it's a mutable struct.
(and I think you overwrite it anyhow most of the time).

How about a default value of `1.0u"eV^-1"`? Then, the pulse is given in units of eV.

fhagemann commented on src/preamp.jl:
```diff
@@ -58,10 +80,16 @@ function GenericPreAmp(preamp_config::PropDict)
+        offset_keV = haskey(preamp_config, :offset_in_keV) ? T(preamp_config.offset_in_keV)u"keV" : 0u"keV",
+        offset_ADC = haskey(preamp_config, :offset_in_ADC) ? T(preamp_config.offset_in_ADC) : 0,
+        # max_e = T(preamp_config.max_e)u"keV",
+        gain = preamp_config.gain/u"eV",
```
The tests fail here because apparently, `preamp_config.gain` is missing.

sagitta42 replied:
Updated the tests, thanks.

oschulz commented:
> Good to merge from my side.

Thanks @fhagemann for looking to this more deeply!

oschulz commented:
We should make this package require Julia v1.9 (we started to do this with other Legend...jl packages already).

fhagemann commented:
> We should make this package require Julia v1.9 (we started to do this with other Legend...jl packages already).

This should become a separate PR. Wanna open one? @oschulz

Commit:
bug fixes, tests updated

sagitta42 commented:
Thanks @fhagemann for the deep look, you found a lot! ^^()

Added basically all your comments: `ustrip` with the unit inside for ensuring, default values for preamp parameters, error check if values in both ADC and keV are provided, updated tests.

fhagemann commented on src/preamp.jl:
```diff
@@ -48,7 +48,7 @@ This struct is currently mutable because in the case of noise modelling based on
     #         typemax(UInt16) / uconvert(u"eV", max_e + offset_keV) # ADC/eV
     #     end            
-    gain::typeof(1.0/1u"eV")
+    gain::typeof(1.0/1u"eV") = 0.0138/u"eV"
```
How did you come up with this value? Is that a typical LEGEND/HADES value?

sagitta42 replied:
This is a value I tuned roughly by eye on the FEP peak for V04545A based on Th228 char data

fhagemann commented on src/preamp.jl:
```diff
@@ -77,17 +77,30 @@ Construct a GenericPreAmp instance based on given simulation configuration <sim_
 function GenericPreAmp(preamp_config::PropDict)
     T = Float32 # This should be somehow defined and be passed properly
+    # check inputs 
+    if haskey(preamp_config, :noise_sigma_keV) && haskey(preamp_config, :noise_sigma_ADC) 
+        if preamp_config.noise_sigma_keV != 0 && preamp_config.noise_sigma_ADC != 0
+            error("You provided preamp noise σ value in both ADC and keV units! Provide only one or the other.")
+        end
+    end
+
+    if haskey(preamp_config, :offset_in_keV) && haskey(preamp_config, :offset_in_ADC)
+        if preamp_config.offset_in_keV != 0 && preamp_config.offset_in_ADC != 0
+            error("You provided preamp offset value in both ADC and keV units! Provide only one or the other.")
+        end
+    end
```
Similar issue as before: `preamp_config.offset_in_keV` is expected to have the unit `u"keV"`, so `preamp_config.offset_in_keV != 0` will **always** return `true`.

I would highly recommend using `iszero` and `!iszero` instead of `== 0` and `!= 0`!

sagitta42 replied:
No, `preamp_config` is just a PropDict based on the Dict/json user provided that has no units in it. Then after these checks a Preamp object is created and the units are added, then the struct itself has units

sagitta42 commented:
Shall we merge @oschulz @fhagemann ?

sagitta42 commented:
Ok after some wrong rebases and pushes and commits and more rebases I figured out what to do. There's the ugly "undo" commit now cause I forgot to undo, I guess let it just be there.

I just wanna merge this so I can run the notebook on NERSC "as if I was the user" with just `add LegendGeSim#main`

fhagemann reviewed (approved):
I got rid of the `undo` commit and it should be good to go now.
